zerver/views/streams.py

```diff
@@ -867,9 +867,20 @@ def delete_in_topic(
         ).values_list("message_id", flat=True)
         messages = messages.filter(id__in=deletable_message_ids)
 
-    messages = messages.select_for_update(of=("self",))
-
-    do_delete_messages(user_profile.realm, messages)
+    # Topics can be large enough that this request will inevitably time out.
+    # In such a case, it's good for some progress to be accomplished, so that
+    # full deletion can be achieved by repeating the request. For that purpose,
+    # we delete messages in atomic batches, committing after each batch.
+    # TODO: Ideally this should be moved to the deferred_work queue.
+    batch_size = RETENTION_STREAM_MESSAGE_BATCH_SIZE
+    while True:
+        with transaction.atomic(durable=True):
+            messages_to_delete = messages.order_by("id")[0:batch_size].select_for_update(
```

Comment:
What do you think about going in reverse order? The thinking being that this would be a bit more readable.

Reply:
Yeah, that should be a much better user experience, i've changed the order.